src/environmentd/src/lib.rs

```diff
@@ -477,15 +477,24 @@ impl Listeners {
                     None,
                 )
                 .await?
+
+            // In read-only mode, we intentionally do not call `set_is_leader`,
+            // because we are by definition not the leader if we are in
+            // read-only mode.
         } else {
-            openable_adapter_storage
+            let adapter_storage = openable_adapter_storage
                 .open(
                     boot_ts,
                     &bootstrap_args,
                     config.controller.deploy_generation,
                     None,
                 )
-                .await?
+                .await?;
+
+            // Once we have successfully opened the adapter storage in
+            // read/write mode, we can announce we are the leader, as we've
+            // fenced out all other environments using the adapter storage.
+            deployment_state.set_is_leader();
         };
 
         // Load the adapter catalog from disk.
```

Comment:
I'm just noticing that this comment seems to have gotten lost and probably belongs before the previous block.

Reply:
Good catch. Updated.